### PR TITLE
Computing state hash as part of transaction processing

### PR DESCRIPTION
**Pull request by jakrawcz-rdx**
As the title says: this PR only contains the "compute state hash based on hash tree update" part.
The resulting state hash is NOT yet stored in database in any way, NOR yet exposed via any API.

### CODE REVIEW

File: core-rust/state-manager/src/store/hash_tree_codec.rs
Comment (jakrawcz-rdx) [reply core-rust/state-manager/src/store/hash_tree_codec.rs:76]: I tried it using enum initially (with `<enum name> = <explicit int tag>` construct), but matching the `type_discriminator: u8` to one of enum values was looking really awkward. Maybe I don't know some rust trick?

File: core-rust/state-manager/src/staging/cache.rs
Comment (jakrawcz-rdx) [reply core-rust/state-manager/src/staging/cache.rs:97]: I see... but that's simply less convenient API for literally 100% of its consumers.

I absolutely hate the `empty_state_diff` field here, but I am lacking some rust idiom - I'd be fine if I could allocate a "statically-owned" singleton with an empty `StateDiff` instance which can be referred to from anywhere. I actually had the same problem here:
https://github.com/radixdlt/babylon-node/blob/3d5627795d1b44231d58c2722dff60fafb234433/core-rust/state-manager/src/staging/stage_tree.rs#L130
Do you see a solution?

File: core-rust/state-manager/src/staging/cache.rs
Comment (dhedey) [inline core-rust/state-manager/src/staging/cache.rs:192]: Very minor, but I'd probably expect all stores to take ownership of the key on insert - so this should possibly be an owned key here in the interface (as that defers the cloning to the parent layer and might be more resistant to unnecessary cloning)

File: core-rust/state-manager/src/staging/cache.rs
Comment (dhedey) [inline core-rust/state-manager/src/staging/cache.rs:179]: Can we give this a better name than `readable_delegate`? Is this actually the root store?

File: core-rust/state-manager/src/staging/cache.rs
Comment (dhedey) [inline core-rust/state-manager/src/staging/cache.rs:121]: Just a thought - but could we call this eg execute_transaction or something? As it's a lazy method now? :)

File: core-rust/state-manager/src/staging/cache.rs
Comment (dhedey) [inline core-rust/state-manager/src/staging/cache.rs:279]: This could return `Option<&StateDiff>` to save storing the `empty_state_diff` - give me a shout if you'd like to pair on this 👍

File: core-rust/state-manager/src/state_manager.rs
Comment (dhedey) [inline core-rust/state-manager/src/state_manager.rs:245]: I'd possibly support calling this `execute_for_staging_with_cache` or something - to distinguish from the mempool execution

**Review by dhedey — COMMENTED** [review]
Nice work! A few misc thoughts - happy to re-review when they're in 👍

File: core-rust/state-manager/src/staging/cache.rs
Comment (jakrawcz-rdx) [reply core-rust/state-manager/src/staging/cache.rs]: @malexandru-rdx plz see the latest commit with the `lazy_init!` that we discussed as an option

Comment (malexandru-rdx) [reply core-rust/state-manager/src/staging/cache.rs]: Definitely better than having 1 extra field. I'm not sure about multithreading implications though (docs says it is behind an atomic lock for `Deref`).

Might need to change to `Option<>` at some point in the future anyway because of this. If nobody else feels strongly about it, I'd leave it like this for now.

File: core-rust/state-manager/src/store/hash_tree_codec.rs
Comment (jakrawcz-rdx) [reply core-rust/state-manager/src/store/hash_tree_codec.rs:132]: As discussed, let's do it: https://github.com/radixdlt/radixdlt-scrypto/pull/806

🛑  This PR is now on hold until the above dep is merged.

Comment (jakrawcz-rdx) [reply core-rust/state-manager/src/store/hash_tree_codec.rs]: 🟢 Unblocked; @dhedey PTAL

**Review by malexandru-rdx — APPROVED** [review]
Great work! :D IMO feel free to merge it.

**Pull request merged**
